pkg/scheduler/queue/queue.go

```diff
@@ -54,229 +54,354 @@ type Request interface{}
+						lastAvailableQuerier = previousQuerier
+					}
+
+					currentQuerier = currentQuerier.next
+				} else {
+					// Nothing for this querier right now, advance through the list.
+					previousQuerier = currentQuerier
+					currentQuerier = currentQuerier.next
+				}
+			}
+		}
+
+		if stopping && (queues.len() == 0 || q.connectedQuerierWorkers.Load() == 0) {
+			// Tell any waiting GetNextRequestForQuerier calls that nothing is coming.
+			for firstAvailableQuerier != nil {
+				firstAvailableQuerier.processed <- nextRequestForQuerier{err: ErrStopped}
+				firstAvailableQuerier = firstAvailableQuerier.next
+			}
+
+			// We are done.
+			close(q.stopCompleted)
+			return
+		}
 	}
+}
 
-	queue := q.queues.getOrAddQueue(userID, maxQueriers)
+func (q *RequestQueue) handleNewRequest(queues *queues, r enqueueRequest) error {
+	queue := queues.getOrAddQueue(r.userID, r.maxQueriers)
 	if queue == nil {
 		// This can only happen if userID is "".
 		return errors.New("no queue found")
 	}
 
 	select {
-	case queue <- req:
-		q.queueLength.WithLabelValues(userID).Inc()
-		q.cond.Broadcast()
-		// Call this function while holding a lock. This guarantees that no querier can fetch the request before function returns.
-		if successFn != nil {
-			successFn()
+	case queue <- r.req:
+		q.queueLength.WithLabelValues(r.userID).Inc()
+		// Call the successFn before sending this request to a waiting querier.
+		if r.successFn != nil {
+			r.successFn()
 		}
 		return nil
 	default:
-		q.discardedRequests.WithLabelValues(userID).Inc()
+		q.discardedRequests.WithLabelValues(r.userID).Inc()
 		return ErrTooManyRequests
 	}
 }
 
-// GetNextRequestForQuerier find next user queue and takes the next request off of it. Will block if there are no requests.
-// By passing user index from previous call of this method, querier guarantees that it iterates over all users fairly.
-// If querier finds that request from the user is already expired, it can get a request for the same user by using UserIndex.ReuseLastUser.
-func (q *RequestQueue) GetNextRequestForQuerier(ctx context.Context, last UserIndex, querierID string) (Request, UserIndex, error) {
-	q.mtx.Lock()
-	defer q.mtx.Unlock()
-
-	querierWait := false
-
-FindQueue:
-	// We need to wait if there are no users, or no pending requests for given querier.
-	for (q.queues.len() == 0 || querierWait) && ctx.Err() == nil && !q.stopped {
-		querierWait = false
-		q.cond.Wait(ctx)
+// dispatchRequestToQuerier finds and forwards a request to a querier, if a suitable request is available.
+// Returns true if this querier should be removed from the list of waiting queriers (eg. because a request has been forwarded to it), false otherwise.
+func (q *RequestQueue) dispatchRequestToQuerier(queues *queues, querier *availableQuerier) bool {
+	if querier.ctx.Err() != nil {
+		querier.processed <- nextRequestForQuerier{
+			next: querier.lastUserIndex,
+			err:  querier.ctx.Err(),
+		}
+		return true
 	}
 
```

Comment:
What's the benefit of tracking the context cancelation here? Can't we only track it in `GetNextRequestForQuerier`?

Reply:
If we track it in `GetNextRequestForQuerier`, there's a potential race between context cancellation and receiving a query that might result in a query being dropped. I've rejigged context cancellation handling in 5944299 to fix an issue where `GetNextRequestForQuerier` gets stuck waiting forever if its context is cancelled but no further requests arrive.